crates/store/src/filesystem.rs

```diff
@@ -0,0 +1,447 @@
+    {
+        let mut keys = vec![];
+
+        if let Ok(entries) = self.root.read_dir(&self.key_root).await {
+            for key_dir in entries {
+                if let Ok(Ok(key_dir)) = key_dir.map(|dir| dir.open_dir()) {
+                    if let Ok(entries) = key_dir.entries().await {
+                        for key_file in entries {
+                            let key_file = match key_file {
+                                Err(_) => continue,
+                                Ok(key_file) => key_file,
+                            };
+
+                            if key_file.file_name().to_string_lossy().contains(KEY_FILE) {
+                                let key_file = key_file.open();
+
+                                keys.push(async {
+                                    read_bincode_file(&mut key_file.map_err(Error::Io)?).await
+                                });
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        join_all(keys.into_iter()).await
+    }
+
+    async fn get_key_dir(&self, expected_identity: &str) -> Result<PathBuf, Error>
+    where
+        K: Serialize + Sync + Send + Eq + Hash,
+    {
+        let key_dir: PathBuf = format!("{}/{}", self.key_root, expected_identity).into();
+
+        if !self.root.exists(&key_dir).await {
+            return Err(Error::NotFound);
+        }
+
+        let mut key_file = key_dir.clone();
+        key_file.push(KEY_FILE);
+
+        let actual_identity = sha256(&read(&mut open(&self.root, &key_file).await?).await?).await?;
+
+        if expected_identity != actual_identity {
+            return Err(Error::InvalidKeyFile);
+        }
+
+        Ok(key_dir)
+    }
+}
+
+pub struct FilesystemCreateItem<'a, K> {
+    fs: &'a mut Filesystem<K>,
+    key_identity: String,
+    mime: Mime,
+    buf: Vec<u8>,
+}
+
+impl<'a, K> AsyncWrite for FilesystemCreateItem<'a, K> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        Pin::new(&mut self.buf).poll_write(cx, buf)
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        Pin::new(&mut self.buf).poll_flush(cx)
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        Pin::new(&mut self.buf).poll_close(cx)
+    }
+}
+
+#[async_trait]
+impl<'a, K> CreateItem for FilesystemCreateItem<'a, K>
+where
+    K: Send + Sync,
+{
+    async fn finish(self) -> (u64, ContentDigest) {
+        let size = self.buf.len() as _;
+
+        // TODO: Compute hash while writing
+        let mut buf: Vec<u8> = Vec::with_capacity(self.buf.len());
+        let mut hasher = Algorithms::default().writer(&mut buf);
+        copy(&mut self.buf.as_slice(), &mut hasher).await.unwrap();
+        let hash = hasher.digests();
```

Comment:
This should use the internal `hash` crate directly, not attempt to compute the content digest set